lib/ansible/utils/collection_loader/_collection_finder.py

```diff
@@ -78,6 +84,141 @@ def is_python_identifier(self):  # type: (str) -> bool
 
 
 PB_EXTENSIONS = ('.yml', '.yaml')
+SYNTHETIC_PACKAGE_NAME = '<ansible_synthetic_collection_package>'
+
+
+class _AnsibleNSTraversable:
+    """Class that implements the ``importlib.resources.abc.Traversable``
+    interface for the following ``ansible_collections`` namespace packages::
+
+    * ``ansible_collections``
+    * ``ansible_collections.<namespace>``
+
+    These namespace packages operate differently from a normal Python
+    namespace package, in that the same namespace can be distributed across
+    multiple directories on the filesystem and still function as a single
+    namespace, such as::
+
+    * ``/usr/share/ansible/collections/ansible_collections/ansible/posix/``
+    * ``/home/user/.ansible/collections/ansible_collections/ansible/windows/``
+
+    This class will mimic the behavior of various ``pathlib.Path`` methods,
+    by combining the results of multiple root paths into the output.
+
+    This class does not do anything to remove duplicate collections from the
+    list, so when traversing either namespace patterns supported by this class,
+    it is possible to have the same collection located in multiple root paths,
+    but precedence rules only use one. When iterating or traversing these
+    package roots, there is the potential to see the same collection in
+    multiple places without indication of which would be used. In such a
+    circumstance, it is best to then call ``importlib.resources.files`` for an
+    individual collection package rather than continuing to traverse from the
+    namespace package.
+
+    Several methods will raise ``NotImplementedError`` as they do not make
+    sense for these namespace packages.
+    """
+    def __init__(self, *paths):
+        self._paths = [pathlib.Path(p) for p in paths]
```

Comment:
Will this cause a `NameError` if there's no `pathlib` per the way you're suppressing the import at the top? If there's a way this is only used under Python 3.5+, it should probably be explained in a code comment. And if it only runs on the controller, I don't see why it's suppressed at all — this would also be a good thing to document.

Reply:
The entire `collection_loader` dir needs to minimally support Py2 due to its use by ansible-test for pytest and import sanity. But since this feature is controller only, and we only support py3 for controller code, we'll not actually run into that issue.

There are a lot of py3 only conditional things already here, it's just common with everything in this directory.